developer_manual/digging_deeper/rest_apis.rst

```diff
@@ -79,3 +81,72 @@ To add an additional method or header or allow less headers, simply pass additio
         }
 
     }
+
+.. _ocs-vs-rest:
+
+Relation of REST and OCS
+------------------------
+
+There is a close relationship between REST APIs and :ref:`OCS <ocscontroller>`.
+Both provide a way to transmit data between the backend of the app in the Nextcloud server and some frontend.
+
+The following combinations of attributes might be relevant for various scenarios:
+
+#. Plain frontend route: ``Controller`` class
+#. Plain Frontend with CRSF checks disabled: ``Controller`` class and ``#[NoCSRFRequired]`` attribute on the method
+#. OCS-based route: ``OCSController`` class
+#. OCS-based route with CSRF disabled: ``OCSController`` class and ``#[NoCSRFRequired]`` attribute on the method
```

Comment:
I think this case doesn't exist, NoCSRFRequired should not have an effect on OCSController, but I will have to check that in the code to be sure.

Reply:
I was thinking back and forth on this as well. In the end, I decided to keep the case as there is a certain difference (at least from the implementation):

For methods with `NoCSRFRequired`, [this section](https://github.com/nextcloud/server/blob/5bd1873511c214330822b6579a1f5374ed7a5dfe/lib/private/AppFramework/Middleware/Security/SecurityMiddleware.php#L186-L207) is doing nothing.

For the case where the attribute is not given, the strict cookie check boils down to a check on [the `OCS-APIREQUEST` header](https://github.com/nextcloud/server/blob/5bd1873511c214330822b6579a1f5374ed7a5dfe/lib/private/AppFramework/Http/Request.php#L454-L457) or no valid session token.

With the header in place (as the OCS controller generally expects), the second part is also skipped: The CSRF check is [hardcoded to true](https://github.com/nextcloud/server/blob/5bd1873511c214330822b6579a1f5374ed7a5dfe/lib/private/AppFramework/Http/Request.php#L430-L432), thus [`isInvalidCSRFRequired`](https://github.com/nextcloud/server/blob/5bd1873511c214330822b6579a1f5374ed7a5dfe/lib/private/AppFramework/Middleware/Security/SecurityMiddleware.php#L226C19-L226C40) returns false.

When the header is not set, you enter the[ second if block](https://github.com/nextcloud/server/blob/5bd1873511c214330822b6579a1f5374ed7a5dfe/lib/private/AppFramework/Middleware/Security/SecurityMiddleware.php#L195-L207) and will check the inner if. The check `isinstance OCSController` is obviously true. The [`isValidOCSRequest`](https://github.com/nextcloud/server/blob/5bd1873511c214330822b6579a1f5374ed7a5dfe/lib/private/AppFramework/Middleware/Security/SecurityMiddleware.php#L234-L237) is somewhat redundant as it can only be reached with the header unset.  Alternatively, a bearer authorization header is sufficient to make the CSRF checker happy (even without OCS header and without `NoCSRFRequired` attribute).

I was adding this case as a reaction on your statement

> There is another case with OCS+CSRF

In my opinion the last case could be dropped (and was never planned in the first round) but I do not see clearly the motivation of it, to be honest. Did I misunderstand your statement above?

Reply:
> In my opinion the last case could be dropped (and was never planned in the first round) but I do not see clearly the motivation of it, to be honest. Did I misunderstand your statement above?

I messed up, I meant OCS+CORS...
Currently the CORS cases are completely missing in the list.

Reply:
Yeah, okay, then that makes more sense. I can read the cors posts and extend by the OCS+cors as well.

I will add a warning about cors in general similar to csrf of you do not mind.

Reply:
If you have any idea how to describe the case *Access from external website* more precisely or otherwise better, feel free to comment :wink:.